drivers/wifi/siwx91x/siwx91x_wifi_socket.c

```diff
@@ -0,0 +1,415 @@
+	 */
+	memcpy(out, in, in_len);
+	if (in->sa_family == AF_INET6) {
+		for (i = 0; i < ARRAY_SIZE(in6->sin6_addr.s6_addr32); i++) {
+			out6->sin6_addr.s6_addr32[i] = ntohl(in6->sin6_addr.s6_addr32[i]);
+		}
+		out6->sin6_port = ntohs(in6->sin6_port);
+	} else if (in->sa_family == AF_INET) {
+		out6->sin6_port = ntohs(in6->sin6_port);
+	}
+}
+
+void siwx91x_on_join_ipv4(struct siwx91x_dev *sidev)
+{
+	sl_net_ip_configuration_t ip_config4 = {
+		.mode = SL_IP_MANAGEMENT_DHCP,
+		.type = SL_IPV4,
+	};
+	struct in_addr addr4 = { };
+	int ret;
+
+	if (!IS_ENABLED(CONFIG_NET_IPV4)) {
+		return;
+	}
+	/* FIXME: support for static IP configuration */
+	ret = sl_si91x_configure_ip_address(&ip_config4, SL_SI91X_WIFI_CLIENT_VAP_ID);
+	if (!ret) {
+		memcpy(addr4.s4_addr, ip_config4.ip.v4.ip_address.bytes, sizeof(addr4.s4_addr));
+		/* FIXME: also report gateway (net_if_ipv4_router_add()) */
+		net_if_ipv4_addr_add(sidev->iface, &addr4, NET_ADDR_DHCP, 0);
+	} else {
+		LOG_ERR("sl_si91x_configure_ip_address(): %#04x", ret);
+	}
+}
+
+void siwx91x_on_join_ipv6(struct siwx91x_dev *sidev)
+{
+	sl_net_ip_configuration_t ip_config6 = {
+		.mode = SL_IP_MANAGEMENT_DHCP,
+		.type = SL_IPV6,
+	};
+	struct in6_addr addr6 = { };
+	int ret, i;
+
+	if (!IS_ENABLED(CONFIG_NET_IPV6)) {
+		return;
+	}
+	/* FIXME: support for static IP configuration */
+	ret = sl_si91x_configure_ip_address(&ip_config6, SL_SI91X_WIFI_CLIENT_VAP_ID);
+	if (!ret) {
+		for (i = 0; i < ARRAY_SIZE(addr6.s6_addr32); i++) {
+			addr6.s6_addr32[i] = ntohl(ip_config6.ip.v6.global_address.value[i]);
+		}
+		/* SiWx91x already take care of DAD and sending ND is not
+		 * supported anyway.
+		 */
+		net_if_flag_set(sidev->iface, NET_IF_IPV6_NO_ND);
+		/* FIXME: also report gateway and link local address */
+		net_if_ipv6_addr_add(sidev->iface, &addr6, NET_ADDR_AUTOCONF, 0);
+	} else {
+		LOG_ERR("sl_si91x_configure_ip_address(): %#04x", ret);
+	}
+}
+
+static int siwx91x_sock_recv_sync(struct net_context *context,
+				  net_context_recv_cb_t cb, void *user_data)
+{
+	struct net_if *iface = net_context_get_iface(context);
+	int sockfd = (int)context->offload_context;
+	struct net_pkt *pkt;
+	struct net_buf *buf;
+	int ret;
+
+	pkt = net_pkt_rx_alloc_on_iface(iface, K_MSEC(100));
+	if (!pkt) {
+		return -ENOBUFS;
+	}
+	buf = net_buf_alloc(&siwx91x_rx_pool, K_MSEC(100));
+	if (!buf) {
+		net_pkt_unref(pkt);
+		return -ENOBUFS;
+	}
+	net_pkt_append_buffer(pkt, buf);
+
+	ret = sl_si91x_recvfrom(sockfd, buf->data, NET_ETH_MTU, 0, NULL, NULL);
+	if (ret < 0) {
+		net_pkt_unref(pkt);
+		ret = -errno;
+	} else {
+		net_buf_add(buf, ret);
+		net_pkt_cursor_init(pkt);
+		ret = 0;
+	}
+	if (cb) {
+		cb(context, pkt, NULL, NULL, ret, user_data);
+	}
+	return ret;
+}
+
+static void siwx91x_sock_on_recv(sl_si91x_fd_set *read_fd, sl_si91x_fd_set *write_fd,
+				 sl_si91x_fd_set *except_fd, int status)
+{
+	/* When CONFIG_NET_SOCKETS_OFFLOAD is set, only one interface exist */
```

Comment:
This is mostly true, but it is possible to have offloaded and non-offloaded interfaces at the same time (at least when considering upper layers, but perhaps not when using this driver).

Reply:
hmm... considering `subsys/net/ip/net_context.c:625`:
```
	/* FIXME - Figure out a way to get the correct network interface
	 * as it is not known at this point yet.
	 */
	if (IS_ENABLED(CONFIG_NET_OFFLOAD) && net_if_is_ip_offloaded(net_if_get_default())) {
		ret = net_offload_get(net_if_get_default(), family, type, proto, context);
```
I have considered that simultaneous offloaded and non-offloaded interfaces was not really supported. Did I miss something?